transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+            return;
+        }
+        else {
+            throwIOException(env, exceptionMessage("Error while submitting IO:", -result));
+            return;
+        }
+    }
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitWrite
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferWrite, jobject callback) {
+
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    struct iocb * iocb = getIOCB(theControl);
+
+    if (iocb == NULL) {
+       throwIOException(env, "Not enough space on the queue for submitting");
+       return;
+    }
+
+    io_prep_pwrite(iocb, fileHandle, getBuffer(env, bufferWrite), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitRead
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferRead, jobject callback) {
+
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    struct iocb * iocb = getIOCB(theControl);
+
+    if (iocb == NULL) {
+       throwIOException(env, "Not enough space on the queue for submitting");
+       return;
+    }
+
+    io_prep_pread(iocb, fileHandle, getBuffer(env, bufferRead), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT jint JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_poll
+  (JNIEnv * env, jobject obj, jobject contextPointer, jobjectArray callbacks, jint min, jint max) {
+
+    int i = 0;
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    int result = io_getevents(theControl->ioContext, min, max, theControl->events, 0);
+
+    for (i = 0; i < result; i++) {
+        struct iocb * iocbp = (struct iocb *)theControl->events[i].obj;
+
+        int eventResult = theControl->events[i].res;
+
+        #ifdef DEBUG
+            fprintf (stderr, "Poll res: %d totalRes=%d\n", eventResult, result);
+        #endif
+
+        if (eventResult < 0) {
+            #ifdef DEBUG
+                fprintf (stderr, "Error: %s\n", strerror(-eventResult));
+            #endif
+
+            jstring jstrError = (*env)->NewStringUTF(env, strerror(-eventResult));
+
+            jobject errorObject = (*env)->NewObject(env,
+                                                   theControl->errorInfoClass,
+                                                   theControl->errorInfoConstr,
+                                                   (jobject)(iocbp->data),
+                                                   (jint)(-eventResult),
+                                                   jstrError);
+
+            if (errorObject == NULL) {
+                return -1;
```

Comment:
Ah, I missed your point about this.

Do we really need to be detailed at that level here?

if this returns NULL, the JNI will have an exception thrown. Meaning: things are really bad at this point.


This would be literally: treat the exception of the exception of the exception.

An error there should be a rare event, Not having the Object available... things are really bad anyways!!!    It can't get any worse!